Oscode/new solver (#3)
SteveBronder/scratch: design_docs/oscode/oscode_sundials.md

The second order ODE can be reposed as a set of two, first order ODEs to fit into the standard schema for sundials solvers,

$$
\dot{\mathbf{y}} = \begin{pmatrix}
\mathbf{y}[1]\\
-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
\end{pmatrix}
$$

Among the SUNDIALS modules, ARKODE ---which solves equations with known explicit and implicit, stiff and nonstill timescale components--- is structured most similarly to OSCODE.

$$
M(t)\dot{\mathbf{y}} = f^E(t, \mathbf{y}) + f^I(t, \mathbf{y}), \quad \mathbf{y}(t_0) = \mathbf{y}_0
$$

For the new OSCODE module we propose the right-hand-side will be a subset of what ARKODE can handle, specifically,

$$
M(t)\dot{\mathbf{y}} = f^O(t, \mathbf{y}), \quad \mathbf{y}|(t_0) = \mathbf{y}_0, \\
$$
where $f^O(t)$ takes the specific form
$$
f^O(t, y) = \begin{pmatrix}
\mathbf{y}[1]\\
-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
\end{pmatrix}.
$$

> As the solution $\mathbf{y}(t)$ may vary between oscillatory and smoothly varying, OSCODE will dynamically switch between a Runge-Kutta solver and its asymptotic ARDC-based solver, at each step choosing the method which yields the larger step length (while keeping the local error within user defined tolerance).

Author: @fruzsinaagocs After this section can you write a blurb describing the ARDC method?

> As the solution $\mathbf{y}(t)$ may vary between oscillatory and smoothly varying, OSCODE will dynamically switch between a Runge-Kutta solver and its asymptotic ARDC-based solver, at each step choosing the method which yields the larger step length (while keeping the local error within user defined tolerance).

Reviewer: defining the acronym here since this is the first mention of it

```suggestion
As the solution $\mathbf{y}(t)$ may vary between oscillatory and smoothly varying, OSCODE will dynamically switch between a Runge-Kutta solver and its asymptotic solver based on the adaptive Riccati defect correction method (ARDC), at each step choosing the method which yields the larger step length (while keeping the local error within user defined tolerance).
```

------------------------------------

(Note: Much of this is taken from the ARKODE docs and modified for OSCODE)

<!--- I removed/modified many of the statements because they didn't apply for OSCODE.--->

The OSCODE infrastructure provides adaptive-step time integration modules for a class of nonstiff Ordinary Differential Equations (ODEs) where the components may be highly oscillatory. Current users of ARKODE will notice a similar structure in the C API of OSCODE.

OSCODE supports ODE systems posed in the form

$$
M(t)\dot{\mathbf{y}} = f^O(t, \mathbf{y}), \quad \mathbf{y}|(t_0) = \mathbf{y}_0, \\
$$
where $f^O(t)$ takes the specific form
$$
f^O(t, y) = \begin{pmatrix}
\mathbf{y}[1]\\
-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
\end{pmatrix}.
$$

Here, $t$ is the independent variable, $\mathbf{y}$ is a set of two (complex-valued) dependent variables with one being the derivative of the other, $\omega$ and $\gamma$ are user specified, real-valued callables. The solver has the ability to switch dynamically between using a Runge-Kutta method and the adaptive Riccati defect correction method to advance the solution, while adaptively updating its stepsize.

> 

Reviewer: The ARDC method approximates the solution with an asymptotic expansion computed numerically, adaptive in the number of terms it retains in the expansion. The approximation is accurate for many wavelengths of an oscillatory solution provided that $\omega$ is large and $\dot{\omega}/\omega$ is small over the interval, thus allowing OSCODE to increase its stepsize while traversing an oscillatory region. These criteria ensure that there exists a nonoscillatory ``phase function'', an underlying slowly changing quantity, which may be approximated well via Chebyshev (or other) polynomials. The ARDC method provides a local error estimate over its timestep, and is capable of interpolating a highly oscillatory solution at little additional cost (around 1 microsecond per output point) by performing interpolation of the phase function.

# Proposal For Including OSCODE Into Sundials

## Summary

> This design document describes the motivation and architecture for adding the OSCODE numerical routine for the efficient solution of one dimensional, second order, linear, homogeneous ordinary differential equations with rapidly oscillating solutions. The proposed solver will will operate with differential equations of the form

Reviewer: one-dimensional is redundant (ordinary already covers that); added ODE abbreviation; state that we solve initial rather than boundary value problems.

```suggestion
This design document describes the motivation and architecture for adding the OSCODE numerical routine for the efficient solution of second order, linear, homogeneous ordinary differential equations (ODEs) with rapidly oscillating solutions. The proposed solver will will operate with initial value problems (IVPs) of the form
```

> $$ \ddot{x} + 2 \gamma(t) \dot{x}(t) + \omega^2(t)x(t) = 0 $$

Author: Leaving a comment here
```suggestion
$$ \ddot{x} + 2 \gamma(t) \dot{x_0}(t) + \omega^2(t)x(t) = 0 $$
```

> $$ \ddot{x} + 2 \gamma(t) \dot{x}(t) + \omega^2(t)x(t) = 0 $$

Reviewer: Suggest $u(t)$ instead of $x(t)$ since $x$ is usually a position -- which is not wrong, but it may be confused for an independent variable. Also added solution range. Also note that eqn need to be puncuated as if it were a word in the sentence.

```suggestion
$$ \ddot{u}(t) + 2 \gamma(t) \dot{u}(t) + \omega^2(t)u(t) = 0, \quad t \in [t_0, t_1], $$
```

Author: Sorry where is the eqn you are referencing here?

Reviewer: Line 7, the one in the substitution.

> 

Reviewer: We need initial conditions to fully define the problem, so add:
```suggestion
$$ u(t_0) = u_0, \quad u(t_1) = u_1, $$
```

> where $\gamma$ and $\omega$ may or may not be expressed as a closed form function of time. We further require $\gamma$ and $\omega$ to be real-valued. If $\omega$ is large, the solution may be highly oscillatory, and standard (polynomial-based) numerical methods will require $\mathcal{O}(\omega)$ timesteps/discretization points. In regimes where $\omega$ is large and smooth, OSCODE exploits an asympotic approximation to reduce this computational cost to $\mathcal{O}(1)$ ($\omega$-independent). In other regimes of the solution interval, OSCODE behaves as a Runge--Kutta solver, and is thus robust to changes in the behavior in the solution from oscillatory to non-oscillatory.

Reviewer: ```suggestion
where $\gamma$ and $\omega$ may or may not be expressed as a closed form function of time. We further require $\gamma$ and $\omega$ to be real-valued and $\omega > 0$. If $\omega$ is large, the solution may be highly oscillatory, and standard (polynomial-based) numerical methods will require $\mathcal{O}(\omega)$ timesteps/discretization points. In regimes where $\omega$ is large and smooth, OSCODE exploits an asympotic approximation to reduce this computational cost to $\mathcal{O}(1)$ ($\omega$-independent). In other regimes of the solution interval, OSCODE behaves as a Runge--Kutta solver, and is thus robust to changes in the behavior in the solution from oscillatory to non-oscillatory.
```

<!--- We should cite the oscode papers for details, and the JOSS paper! --->

> This new module will follow the same C based API with an underlying C++ implementation.

Reviewer: I need clarification here. Same API as what? oscode's?

Author: Need to scratch and fix this. Going to add this to the Guide Level Explanation section

```
Current users of ARKODE will notice a similar structure in the C API of OSCODE.
```

tmk we should have a very similar structure since we will probably use the explicit RK solver from ARKODE

Reviewer: OK!

## Motivation

> Solutions to oscillatory problems are often bespoke and use analytic approximations such as the Adaptive Ricatti Defect Correction (ARDC) method used by OSCODE. Giving researchers access to a general purpose solver for solutions for ODEs of the form written above will allow them to explore more computationally complex models in their research and speed up existing implementations.

Reviewer: Ah. So oscode doesn't use the adaptive Riccati defect correction (ARDC) method, it uses the Wentzel--Kramers--Brillouin approximation (WKB). The ARDC is used by the riccati solver.

Author: Ooops yes this was a find and replace error. Fixed!

The second order ODE can be reposed as a set of two, first order ODEs to fit into the standard schema for sundials solvers,

$$
\dot{y} = \begin{pmatrix}
y[1]\\

> -y[0]w^2(t)-2\gamma(t)y[1]

Reviewer: ```suggestion
-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
```

Author: Same for $y[1]$?

> -y[0]w^2(t)-2\gamma(t)y[1]

Reviewer: ```suggestion
\dot{\mathbf{y}} = \begin{pmatrix}
\mathbf{y}[1]\\
-\mathbf{y}[0]w^2(t)-2\gamma(t)\mathbf{y}[1]
```

\end{pmatrix}.
$$

> The most similar module is ARKODE, which solves equations with known explicit and implicit, nonstiff and stiff time scale components,

Reviewer: ```suggestion
Among the SUNDIALS modules, ARKODE ---which solves equations with known explicit and implicit, stiff and nonstill timescale components--- is structured most similarly to OSCODE.
The most similar module is ARKODE, which solves equations with known explicit and implicit, nonstiff and stiff time scale components,
```

$$

> M(t)\dot{y} = f^E(t, y) + f^I(t, y), \quad y(t_0) = y_0.

Reviewer: Bolded dependent variables to make it clear that this is a linear system of ODEs.

```suggestion
M(t)\dot{\mathbf{y}} = f^E(t, \mathbf{y}) + f^I(t, \mathbf{y}), \quad \mathbf{y}(t_0) = \mathbf{y}_0.
```

$$

For the new OSCODE module we propose the right-hand-side will be a subset of what ARKODE can handle, specifically,

$$
M(t)\ddot{y} = f^O(t, y), \quad y(t_0) = y_0 \\
$$

$$
f^O(t, y) = \begin{pmatrix}
y[1]\\
-y[0]w^2(t)-2\gamma(t)y[1]
\end{pmatrix}.

> $$

Reviewer: ```suggestion
$$
M(t)\dot{\mathbf{y}} = f^O(t, \mathbf{y}), \quad \mathbf{y|(t_0) = \mathbf{y}_0, \\
$$

where $f^O(t)$ takes the specific form

$$
f^O(t, y) = \begin{pmatrix}
\mathbf{y}[1]\\
-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
\end{pmatrix}.
$$
```

> Here, $f^O$ is a function with varying sized oscillatory regions. OSCODE will then dynamically switch between a Runge-Kutta solver and its asymptotic ARDC-based solver, at each step choosing the method which yields the larger step length (while keeping the local error within user tolerance).

Reviewer: ```suggestion
As the solution $\mathbf{y}(t)$ may vary between oscillatory and smoothly varying, OSCODE will dynamically switch between a Runge-Kutta solver and its asymptotic ARDC-based solver, at each step choosing the method which yields the larger step length (while keeping the local error within user tolerance).
```

## Guide-level Explanation

------------------------------------
Section Goals:

Explain the proposal as if it was already included in the project and you were teaching it to another Sundials programmer in the manual. That generally means:

- Introducing new named concepts.
- Explaining the feature largely in terms of examples.
- Explaining how Sundials programmers should *think* about the feature, and how it should impact the way they use the relevant package. It should explain the impact as concretely as possible.
- If applicable, provide sample error messages, deprecation warnings, or migration guidance.

<!--- Should the above be deleted? --->
------------------------------------

(Note: Much of this is taken from the ARKODE docs and modified for OSCODE)

<!--- I removed/modified many of the statements because they didn't apply for OSCODE.--->

> The OSCODE infrastructure provides adaptive-step time integration modules for a class of nonstiff ordinary differential equations (ODEs) where the components may be highly oscillatory.  

Reviewer: ```suggestion
The OSCODE infrastructure provides adaptive-step time integration modules for a class of nonstiff ordinary differential equations (ODEs) where the solution may become highly oscillatory.  
```

> OSCODE supports ODE systems posed in the form below

Reviewer: ```suggestion
OSCODE supports ODE systems posed in the form
```

$$
M(t)\ddot{y} = f^O(t, y), \quad y(t_0) = y_0
$$

with

$$
f^O(t, y) = \begin{pmatrix}
y[1]\\
-y[0]w^2(t)-2\gamma(t)y[1]
\end{pmatrix}.

> $$

Reviewer: Please copy and paste my correction from earlier

> Here, $f^O$ is a function with varying sized oscillatory regions. $t$ is the independent variable, $y$ is a set of two (complex-valued) dependent variables with one being the derivative of the other, $\omega$ and $\gamma$ are user specified, real-valued callables. The solver has the ability to switch dynamically between using a Runge-Kutta method and a Wentzel-Kramers-Brillouin method to advance the solution, while adaptively updating its stepsize.

Reviewer: ```suggestion
Here, $t$ is the independent variable, $\mathbf{y}$ is a set of two (complex-valued) dependent variables with one being the derivative of the other, $\omega$ and $\gamma$ are user specified, real-valued callables. The solver has the ability to switch dynamically between using a Runge-Kutta method and the adaptive Riccati defect correction method to advance the solution, while adaptively updating its stepsize.
```

### Adaptive single-step methods

> The OSCODE framework is designed to support single step, IVP integration methods

Reviewer: ```suggestion
The OSCODE framework is designed to support single step, IVP integration methods.
```

$$
\ddot{y} = f(t ,y), y(t_0) = y_0

> $$

Reviewer: cut -- too general, oscode can't actually deal with any equation of this form.

> The choice of step size is determined by the time-stepping method (based on user-provided inputs, typically accuracy requirements). However, users may place minimum/maximum bounds on if desired.

Reviewer: ```suggestion
The stepsize is determined by the time-stepping method (based on user-provided accuracy requirements). However, users may place minimum/maximum bounds on if desired.
```

> At each step, either the ARDC or RK solver is chosen for that step based on the relative difference in error between the two methods. For RK, the error is generated by the difference in the $N$th and $N-1$th order RK steps. The ARDC steps error for the function evaluation and the first derivative of $x$ with respect to time is defined as

Reviewer: ```suggestion
At each step, either the ARDC or RK solver is chosen for that step based on their estimated local error. For RK, the error is generated by the difference in the $N$th and $N-1$th order RK steps. The ARDC step's error is calculated as the residual of a transformed form of the original ODE, by substituting in ARDC's solution estimate.
```

$$
\delta x_{ARDC} = A_+ \delta f_+ + A_- \delta f_-
$$
$$
\delta f_\pm = f_\pm \sum_{i=0}^n \delta[S_i]_t^{t + h}
$$

$$
\delta \dot{x}\_{ARDC} = B\_+ \delta \dot{f}\_+ + B\_- \delta \dot{f}\_-
$$

$$
\delta \dot{f}\_\pm = \delta f\_\pm \frac{\dot{f}\_\pm}{f\_\pm}

> $$

Reviewer: This is actually oscode's WKB step error estimate, not ARDC's! Cut. I suggested a words-only explanation above, the math may beyond the scope of this document.

Author: Yeah I agree with this

OSCODE's time stepping modules may be run in a variety of "modes":

> - NORMAL: The solver will take internal steps until it has just overtaken a user-specified output time, $t_{out}$, in the direction of integration, i.e. $t_{n-1} < t_{out} \le t_n$ for forward integration, or $t_{n} < t_{out} \le t_{n-1}$ for backward integration. It will then compute an approximation to the solution $y(t_{out})$ by interpolation (using one of the dense output routines described in the section TBD).

Reviewer: ```suggestion
- NORMAL: The solver will take internal steps until it has just overtaken a user-specified end time, $t_1$, in the direction of integration, i.e. $t_{n-1} < t_1 \le t_n$ for forward integration, or $t_{n} < t_1 \le t_{n-1}$ for backward integration. It will then compute an approximation to the solution $\mathbf{y}(t_1)$ by interpolation (using one of the dense output routines described in the section TBD).
```

Reviewer: Also, are we planning a section on dense output?

Author: Yes let me go through sundials docs on how they currently phrase dense output information and I'll put that here as well

> - ONE-STEP: The solver will only take a single internal step $y_{n-1} \rightarrow y_n$ and then return control back to the calling program. If this step will overtake $t_{out}$ then the solver will again return an interpolated result; otherwise it will return a copy of the internal solution $y_n$

Reviewer: Nice explanation, I just updated the notation for consistency.

```suggestion
- ONE-STEP: The solver will only take a single internal step $\mathbf{y}_{n-1} \rightarrow \mathbf{y}_n$ and then return control back to the calling program. If this step will overtake $t_1$ then the solver will again return an interpolated result; otherwise it will return a copy of the internal solution $\mathbf{y}_n$.
```

> - NORMAL-TSTOP: The solver will take internal steps until the next step will overtake $t_{out}$. It will then limit this next step so that $t_n = t_{n-1} + h_n = t_{out}$, and once the step completes it will return a copy of the internal solution

Reviewer: ```suggestion
- NORMAL-TSTOP: The solver will take internal steps until the point where the _next_ step would overtake $t_1$. It will then limit this next step so that $t_n = t_{n-1} + h_n = t_1$, and once the step completes it will return a copy of the internal solution.
```

> - ONE-STEP-TSTOP: The solver will check whether the next step will overtake $t_{out}$ if not then this mode is identical to "one-step" above; otherwise it will limit this next step so that $t_n = t_{n-1} + h_n = t_{out}$. In either case, once the step completes it will return a copy of the internal solution $y_n$.

Reviewer: ```suggestion
- ONE-STEP-TSTOP: The solver will check whether the next step would overtake $t_1$. If not, then this mode is identical to "one-step" above; otherwise it will limit this next step so that $t_n = t_{n-1} + h_n = t_1$. In either case, once the step completes it will return a copy of the internal solution $\mathbf{y}_n$.
```

.

> We note that interpolated solutions may be slightly less accurate than the internal solutions produced by the solver. Hence, to ensure that the returned value has full method accuracy one of the "tstp" modes may be used.

Reviewer: Cut, this is just not true for ARDC -- the interpolation's order of accuracy is the same as that of the internal steps'.

> ### Interpolation

Reviewer: Ah, here's the dense output section. Update the link pointing to this above?

> As mentioned above, the time-stepping modules in OSCODE support interpolation of solution $y(t_{out})$ and derivatives $y^{(d)}(t_{out})$, where $t_{out}$ occurs within a completed time step from $t_{n-1} \rightarrow t_n$. Additionally, this module supports extrapolation of solutions and derivatives for outside this interval (e.g. to construct predictors for iterative nonlinear and linear solvers). To this end, OSCODE currently supports construction of polynomial interpolants $p_q(t)$ of polynomial degree up to $q=5$, although users may select interpolants of lower degree.

Reviewer: The order of interpolation is going to be determined by different factors depending on which method the step we're interpolation within was taken with. E.g. If it's a RK step, it'll almost certainly be the same order (or one lower) than that of the RK solver. If it was taken with ARDC, it's set freely by the user, but the same order will be used to perform derivatives, integration within ARDC steps. Maybe this is just a bit too complicated for this white paper.

```suggestion
As mentioned above, the time-stepping modules in OSCODE support interpolation of solution $\mathbf{y}(t_{\mathrm{out}})$ and derivatives $\frac{\mathrm{d}\mathbf{y}}{\mathrm{d}t}|_{t = t_{\mathrm{out}}}$, where $t_{\mathrm{out}}$ occurs within a completed time step from $[t_{n-1},  t_n]$. To this end, OSCODE currently supports construction of polynomial interpolants $p_q(t)$ where the degree $q$ is determined by a parameter passed to the solver by the user.
```

> OSCODE provides three complementary interpolation approaches, both of which are accessible from any of the time-stepping modules: "Hermite" and "Lagrange". The Hermite approach has been included with ARKODE since its inception, and is more suitable for non-stiff problems; Lagrange is designed to provide increased accuracy when integrating stiff problems. These two methods are described in detail within the ARKODE documentation (link).

Reviewer: Was this a typo or is there a third option?

```suggestion
OSCODE provides two complementary interpolation approaches, both of which are accessible from any of the time-stepping modules: "Hermite" and "Lagrange". The Hermite approach has been included with ARKODE since its inception, and is more suitable for non-stiff problems; Lagrange is designed to provide increased accuracy when integrating stiff problems. These two methods are described in detail within the ARKODE documentation (link).
```

  y(tout).  In the OSC_ONE_STEP mode, it takes one internal step
  and returns.  The behavior of both modes can be over-rided
  through user-specification of osc_tstop (through the
  *StepSetStopTime function), in which case if a solver step
  would pass tstop, the step is shortened so that it stops at
  exactly the specified stop time, and hence interpolation of
  y(tout) is not required.
  ---------------------------------------------------------------*/
int oscEvolve(OSCodeMem osc_mem, realtype tout, N_Vector yout,
              realtype *tret, int itask);
```

The `OSCodeMemRec` will be similar to `ARKodeMemRec` containing the information for the estimation using the rk and wkb solver

(TODO:) The example program `example_oscode.cpp` shows an example of the user API for the burst equation.

Users can set:

1. `w` and `g` functions that take in a timestep `t` along with a `void*` argument for their own data
2. `OSCODEOmegaFn` and `OSCODEGammaFn`, the functors $\omega(t)$ and $\gamma(t)$ which have pointer type
    - `complextype (OSCODEOmegaFn*)(realtype)`
3. Options for the rk solver
    - order
    - tolerances
    - butcher tables

>     - Gauss - Lobatto weights

Reviewer: Gauss--Lobatto weights not needed!

>     - Integral Function

Reviewer: What's this?

Author: Scratch that, I was thinking of `g` but we document that earlier

> 4. Options for the wkb solver
>     - Order
>     - order related weight matrices
>     - Derivative and series functions
>     - Function for Hermite and Lagrange
>     - Gauss-Lobatto weights

Reviewer: We won't have a WKB solver. Suggest change to:
```suggestion
4. Options for the ARDC solver
    - Order
    - tolerance(s)
```

5. Options for solver
    - initial conditions for the ODE, $ x(t)  \frac{dx}{dt} $ evaluated at the start of the integration range
    - start of integration range
    - end of integration range
    - do_times timepoints at which dense output is to be produced. Must be sorted

>     - Direction

Reviewer: ```suggestion
    - Direction of integration
```

    - order of ARDC approximation to be used
    - (local) relative tolerance
    - (local) absolute tolerance

>     - initial stepsize to use

Reviewer: ```suggestion
    - initial stepsize suggestion (to be refined by the solver)
```

> The implimentation can utilize the ARKODE explicit runga kutta solver, but we will have to write an implementation of the ARDC solver. For a first implementation I think it would be fine to only have the 6th order ARDC solver already written in C++, then in another PR expose functions that allow the user to change the ARDC solvers order.

Reviewer: ARDC is not fixed order, it's adaptive. Also it's not written in C++ -- not sure what happened here. It'll need to be translated from Python, but it's conceptually simple.

Author: Whoops yes, leftover from WKB

Tests and example code will be written for the airy and burst examples.

## Drawbacks

The main question is the user base. As long as we think there enough researchers that would use this method then it most likely makes sense to add.

The other main question to this proposal is whether it's possible to make this fit in the ARKODE framework.

## Rationale and alternatives

------------------------------------

- Why is this design the best in the space of possible designs?
- What other designs have been considered and what is the rationale for not choosing them?
- What is the impact of not doing this?

------------------------------------
## Prior art

Discuss prior art, both the good and the bad, in relation to this proposal.
A few examples of what this can include are:

- For language, library, tools, and compiler proposals: Does this feature exist in other programming languages and what experience have their community had?

> The OSCODE solver is currently available as a python package and standalone header files.

Reviewer: Add: The RICCATI solver is available as a (pure) python package.

## Papers

Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.

- Original paper: ["Efficient method for solving highly oscillatory ordinary differential equations
with applications to physical systems"](https://journals.aps.org/prresearch/pdf/10.1103/PhysRevResearch.2.013030)
- JOSS Paper on Python / C++ implementation: [Link](https://joss.theoj.org/papers/10.21105/joss.02830)
- C++ Implementation: [Link](https://github.com/fruzsinaagocs/oscode)
- Python Wrapper Docs: [Link](https://oscode.readthedocs.io/en/latest/)

> 

Reviewer: ```suggestion
- Riccati paper: ["An adaptive spectral method for oscillatory second-order linear ODEs with frequency-independent cost"](https://arxiv.org/abs/2212.06924)
- JOSS paper on its Python implementation: [Link](https://joss.theoj.org/papers/10.21105/joss.05430)
```

## Unresolved Questions

- What parts of the design do you expect to resolve through the RFC process before this gets merged?

Whether OSCODE should be its own seperate module or whether it is possible to fit it in the ARKODE framework.

- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?

> I'd like to resolve what the user should have access to modify, for example the butcher tables in the ARDC method.

Reviewer: ARDC doesn't have Butcher tableaus...

Author: Whoops, that's leftover from WKB
